src/Domain/CoordinateMaps/SphericalToCartesianPfaffian.cpp

```diff
@@ -0,0 +1,135 @@
+// See LICENSE.txt for details.
+
+#include "Domain/CoordinateMaps/SphericalToCartesianPfaffian.hpp"
+
+#include <cmath>
+#include <cstddef>
+#include <pup.h>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/DereferenceWrapper.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+namespace domain::CoordinateMaps {
+SphericalToCartesianPfaffian::SphericalToCartesianPfaffian() = default;
+SphericalToCartesianPfaffian::SphericalToCartesianPfaffian(
+    SphericalToCartesianPfaffian&&) = default;
+SphericalToCartesianPfaffian::SphericalToCartesianPfaffian(
+    const SphericalToCartesianPfaffian&) = default;
+SphericalToCartesianPfaffian& SphericalToCartesianPfaffian::operator=(
+    const SphericalToCartesianPfaffian&) = default;
+SphericalToCartesianPfaffian& SphericalToCartesianPfaffian::operator=(
+    SphericalToCartesianPfaffian&&) = default;
+
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3>
+SphericalToCartesianPfaffian::operator()(
+    const std::array<T, 3>& source_coords) const {
+  const auto& [r, theta, phi] = source_coords;
+  return {
+      {r * sin(theta) * cos(phi), r * sin(theta) * sin(phi), r * cos(theta)}};
+}
+
+// NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+std::optional<std::array<double, 3>> SphericalToCartesianPfaffian::inverse(
+    const std::array<double, 3>& target_coords) const {
+  const auto& [x, y, z] = target_coords;
+  if (UNLIKELY(y == 0.0 and x == 0.0)) {
+    if (UNLIKELY(z == 0.0)) {
+      return std::array{0.0, 0.5 * M_PI, 0.0};
+    } else {
+      return std::array{std::abs(z), z > 0.0 ? 0.0 : M_PI, 0.0};
+    }
+  } else {
+    const double r = std::hypot(x, y, z);
+    return std::array{r, acos(z / r),
+                      x > 0.0 ? atan2(y, x) : atan2(y, x) + 2.0 * M_PI};
+  }
+}
+
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>
+SphericalToCartesianPfaffian::jacobian(
+    const std::array<T, 3>& source_coords) const {
+  const auto& [r, theta, phi] = source_coords;
+  using DataType = tt::remove_cvref_wrap_t<T>;
+  tnsr::Ij<DataType, 3, Frame::NoFrame> jacobian_matrix{
+      make_with_value<DataType>(dereference_wrapper(r), 0.0)};
+  const DataType cos_theta = cos(theta);
+  const DataType sin_theta = sin(theta);
+  const DataType cos_phi = cos(phi);
+  const DataType sin_phi = sin(phi);
+  get<0, 0>(jacobian_matrix) = sin_theta * cos_phi;
+  get<1, 0>(jacobian_matrix) = sin_theta * sin_phi;
+  get<2, 0>(jacobian_matrix) = cos_theta;
+  get<0, 1>(jacobian_matrix) = r * cos_theta * cos_phi;
+  get<1, 1>(jacobian_matrix) = r * cos_theta * sin_phi;
+  get<2, 1>(jacobian_matrix) = -r * sin_theta;
+  // Pfaffian basis means phi components are 1 / sin_theta times those of a
+  // coord basis
+  get<0, 2>(jacobian_matrix) = -r * sin_phi;
+  get<1, 2>(jacobian_matrix) = r * cos_phi;
+  // get<2, 2>(jacobian_matrix) is zero
```

Comment:
Now you have extra allocations 🙃 
```cpp
const auto& sin_theta = get<2, 1>(jacobian_matrix) = sin(theta);
const auto& cos_theta = get<2, 0>(jacobian_matrix) = cos(theta);
const auto& sin_phi = get<0, 2>(jacobian_matrix) = sin(phi);
const auto& cos_phi = get<1, 2>(jacobian_matrix) = cos(phi);

get<0, 0>(jacobian_matrix) = sin_theta * cos_phi;
get<1, 0>(jacobian_matrix) = sin_theta * sin_phi;
get<0, 1>(jacobian_matrix) = r * cos_theta * cos_phi;
get<1, 1>(jacobian_matrix) = r * cos_theta * sin_phi;

get<2, 1>(jacobian_matrix) *= -r;
get<0, 2>(jacobian_matrix) *= -r;
get<1, 2>(jacobian_matrix) *= r;
```
This has zero extra allocations and less trig.